src/main/java/hudson/remoting/RemoteInvocationHandler.java

```diff
@@ -410,31 +418,40 @@ public void run() {
             inQueue.set(false); // we have started execution and running is true, so queued can be reset
             try {
                 long nextSweep = System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(200);
+                int count = 0;
                 while (!referenceLists.isEmpty()) {
                     try {
-                        Reference<?> ref = queue.remove(100);
-                        if (ref instanceof PhantomReferenceImpl) {
-                            PhantomReferenceImpl r = (PhantomReferenceImpl) ref;
-                            final Channel.Ref channelRef = r.channel;
-                            try {
-                                r.cleanup();
-                            } catch (IOException e) {
-                                logger.log(Level.WARNING,
-                                        String.format("Couldn't clean up oid=%d from %s", r.oid, r.origin), e);
-                            } catch (Error e) {
-                                logger.log(Level.SEVERE,
-                                        String.format("Couldn't clean up oid=%d from %s", r.oid, r.origin), e);
-                                throw e; // pass on as there is nothing we can do with an error
-                            } catch (Throwable e) {
-                                logger.log(Level.WARNING,
-                                        String.format("Couldn't clean up oid=%d from %s", r.oid, r.origin), e);
-                            } finally {
-                                if (channelRef != null) {
-                                    final List<PhantomReferenceImpl> referenceList = referenceLists.get(channelRef);
-                                    if (referenceList != null) {
-                                        referenceList.remove(r);
-                                        if (channelRef.channel() == null) {
-                                            cleanList(referenceList);
+                        long drainUntil = System.currentTimeMillis() + 200;
```

Comment:
Yes that was a second order change I have in mind. I also want to use an exponentially decaying average rate of unexports per second rather than unexports per sweep to control logging... And maybe have the logging once every 5 minutes

Sent from my iPad

> On 13 Apr 2016, at 23:16, James Nord <notifications@github.com> wrote:
> 
> In src/main/java/hudson/remoting/RemoteInvocationHandler.java:
> 
> > -                                        String.format("Couldn't clean up oid=%d from %s", r.oid, r.origin), e);
> > -                            } catch (Error e) {
> > -                                logger.log(Level.SEVERE,
> > -                                        String.format("Couldn't clean up oid=%d from %s", r.oid, r.origin), e);
> > -                                throw e; // pass on as there is nothing we can do with an error
> > -                            } catch (Throwable e) {
> > -                                logger.log(Level.WARNING,
> > -                                        String.format("Couldn't clean up oid=%d from %s", r.oid, r.origin), e);
> > -                            } finally {
> > -                                if (channelRef != null) {
> > -                                    final List<PhantomReferenceImpl> referenceList = referenceLists.get(channelRef);
> > -                                    if (referenceList != null) {
> > -                                        referenceList.remove(r);
> > -                                        if (channelRef.channel() == null) {
> > -                                            cleanList(referenceList);
> > +                        long drainUntil = System.currentTimeMillis() + 200;
> should keep to nanoTime() to avoid clock jump/resync issues
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> 
> -- 
> You received this message because you are subscribed to the Google Groups "engineering-code-reviews" group.
> To unsubscribe from this group and stop receiving emails from it, send an email to engineering-code-reviews+unsubscribe@cloudbees.com.
> To post to this group, send email to engineering-code-reviews@cloudbees.com.
> To view this discussion on the web visit https://groups.google.com/a/cloudbees.com/d/msgid/engineering-code-reviews/jenkinsci/remoting/pull/81/r59635696%40github.com.